### PR TITLE
Bug 1937594: Split SDN migration into 2 phase

### PR DESCRIPTION
**Pull request by pliurh**
Phase1, Users add the spec.migration feild in network.operator
CR. The only changes the migration field under the status of
Network.config. It will trigger MCO to provision ovs-configuration
service on every node and reboot. After rebooting, openshift-sdn
will use br-ex as L3 gateway.

Phase2, Users update the spec.networkType in network.config.
CNO starts to swap the network provider for the cluster to ovnkube.
Users need to manually reboot all nodes to make all pods attach to
the new cluster network.

### CODE REVIEW

**Comment by pliurh** [comment]
@trozet @abhat @squeed PTAL.

**Comment by pliurh** [comment]
This PR depends on the https://github.com/openshift/machine-config-operator/pull/2015

**Comment by pliurh** [comment]
/unhold
Since there will be no intermediary state (aka old ovn local gateway mode) available anymore. We have to take this approach.

**Comment by pliurh** [comment]
@dcbw @trozet @abhat  PTAL

**Comment by trozet** [comment]
/hold
I think you may not want to split into 2 phases after these go in:
https://github.com/openshift/cluster-network-operator/pull/801
https://github.com/openshift/ovn-kubernetes/pull/281

**Comment by pliurh** [comment]
> /hold
> I think you may not want to split into 2 phases after these go in:
> #801
> [openshift/ovn-kubernetes#281](https://github.com/openshift/ovn-kubernetes/pull/281)

In 4.6, maybe yes. In long term, if the ovn-kube local gw mode will exist until openshift-sdn fades away in OCP, we can always relay on it as the intermediate stage during migration. But, if that is not the case, we will still need this 2-steps approach when the local gw mode is deprecated by ovn-kube.

**Comment by trozet** [comment]
OK I'll leave it to your judgement. I just wanted to make you aware of the other changes before this merged so you could make the call.

/hold cancel

**Comment by pliurh** [comment]
/hold

I'll test the local gw approach first with @trozet 's PRs. If it works, we will go that way in 4.6. And this PR can wait until 4.7.

**Comment by pliurh** [comment]
/unhold
@trozet After testing with https://github.com/ovn-org/ovn-kubernetes/pull/2055, I found the gateway_mode_flags `--gateway-mode local --gateway-interface none` doesn't work anymore due to https://github.com/openshift/ovn-kubernetes/commit/1ba1ce885f089a023b4c803d85a2ffc7206eb98b. It breaks the local gateway mode.
```
I0312 08:00:41.404433   14571 gateway_init.go:162] Initializing Gateway Functionality
I0312 08:00:41.404649   14571 gateway_localnet.go:184] Node local addresses initialized to: map[10.131.0.2:{10.131.0.0 fffffe00} 127.0.0.1:{127.0.0.0 ff000000} 172.22.0.99:{172.22.0.0 ffffff00} 192.168.111.23:{192.168.111.0 ffffff00} 192.168.111.4:{192.168.111.4 ffffffff} ::1:{::1 ffffffffffffffffffffffffffffffff} fe80::899a:2541:edca:cbe8:{fe80:: ffffffffffffffff0000000000000000} fe80::d49c:f8ff:fe3a:98c8:{fe80:: ffffffffffffffff0000000000000000} fe80::dacf:b57b:2ab4:6b96:{fe80:: ffffffffffffffff0000000000000000} fe80::e07b:72ff:fe1b:c0a5:{fe80:: ffffffffffffffff0000000000000000}]
I0312 08:00:41.404820   14571 helper_linux.go:73] Found default gateway interface enp2s0 192.168.111.1
F0312 08:00:41.404860   14571 ovnkube.go:130] could not find IP addresses: failed to lookup link none: Link not found
```

Instead of spending time to fix the local gw mode in ovn-kube, I think it's better to refactor the migration with this PR, so that we don't need to rely on the ovn-kube intermediate stage anymore.

**Comment by pliurh** [comment]
New CI test for this change https://github.com/openshift/release/pull/16861

**Comment by alexanderConstantinescu** [comment]
/retitle Bug 1937594: Split SDN migration into 2 phase

File: pkg/controller/operconfig/operconfig_controller.go
Comment (trozet) [inline pkg/controller/operconfig/operconfig_controller.go:179]: How does this work? If the annotation doesn't exist how do you know it is in network migration mode? To me it would seem the opposite would be true. Should this be

```
if val, ok := operConfig.GetAnnotations()[names.NetworkMigrationAnnotation]; ok && val != "start" {
```

File: pkg/controller/operconfig/operconfig_controller.go
Comment (trozet) [inline pkg/controller/operconfig/operconfig_controller.go:181]: what if val is not prepare? should we log some error saying invalid value for migration?

File: pkg/controller/operconfig/operconfig_controller.go
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go:179]: The condition here is to check if CNO shall bypass `IsChangeSafe`. So if the annotation doesn't exist or it exists but its value is not 'start', both mean CNO shall not bypass `IsChangeSafe` check and start the migration.

If the annotation exists and its value is 'prepare', the CNO comes to the migration preparing phase.

File: pkg/controller/operconfig/operconfig_controller.go
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go:187]: fixed.

File: pkg/controller/operconfig/operconfig_controller.go
Comment (trozet) [reply pkg/controller/operconfig/operconfig_controller.go]: OK I don't think I understand CNO code well enough to review this. @abhat do you mind taking a look?

File: pkg/controller/operconfig/operconfig_controller.go
Comment (abhat) [inline pkg/controller/operconfig/operconfig_controller.go:185]: if `!ok`, why do we want to continue checking for `"prepare"`?

Comment (abhat) [reply pkg/controller/operconfig/operconfig_controller.go:185]: is it better to just say `if ok && val == "prepare" {`

File: pkg/controller/operconfig/operconfig_controller.go
Comment (abhat) [inline pkg/controller/operconfig/operconfig_controller.go:191]: nit: log should be consistent with the status message we are setting?

File: pkg/controller/operconfig/operconfig_controller.go
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go]: fixed.

File: pkg/controller/operconfig/operconfig_controller.go
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go:191]: Actually, these lines are copied from https://github.com/openshift/cluster-network-operator/blob/eb90c4ed641f302f5c1b5d514257c7f0698de66a/pkg/controller/operconfig/operconfig_controller.go#L327

And I found the log and the status messages aren't consistent in other existing codes. I'm not sure whether it's a protocol we need to follow.

**Comment by trozet** [comment]
@danwinship

File: pkg/network/cluster_config.go
Comment (danwinship) [inline pkg/network/cluster_config.go:142]: We can't do this. This is a public API; arbitrary components may be relying on the value of this field, for arbitrary purposes, and they might break if it is suddenly unset.

We may need to add a field to the config indicating that migration is in progress, so MCO can respond to that. (Maybe MCO could just read the annotation? That seems kind of bad though. If it's an actual API between CNO and MCO, then it should be actual API, not an annotation.)

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/network/cluster_config.go
Comment (pliurh) [reply pkg/network/cluster_config.go:142]: When the migration starts, the cluster will be in a maintenance state. The cluster shall be considered as out-of-service until the whole process finishes. We don't guarantee any of the cluster components can behave as normal during this period. So as long as the MCO can work as expected, I suppose it should be fine. WDYT?

**Comment by pliurh** [comment]
Requires MCO change in https://github.com/openshift/machine-config-operator/pull/2518

**Comment by pliurh** [comment]
@danwinship PTAL

File: go.mod
Comment (pliurh) [reply go.mod:19]: https://github.com/openshift/cluster-network-operator/pull/1057

File: pkg/controller/operconfig/operconfig_controller.go
Comment (pliurh) [reply pkg/controller/operconfig/operconfig_controller.go]: fixed.

**Pull request merged**
